allennlp/models/encoder_decoders/copynet_seq2seq.py

```diff
@@ -0,0 +1,824 @@
+        # the weights for the selective read are simply the predicted probabilities
+        # corresponding to each token in the source sentence from the previous timestep.
+        self._target_embedder = Embedding(target_vocab_size, target_embedding_dim)
+        self._attention = attention
+        self._input_projection_layer = Linear(
+                target_embedding_dim + self.encoder_output_dim * 2,
+                self.decoder_input_dim)
+
+        # We then run the projected decoder input through an LSTM cell to produce
+        # the next hidden state.
+        self._decoder_cell = LSTMCell(self.decoder_input_dim, self.decoder_output_dim)
+
+        # We create a "generation" score for each token in the target vocab
+        # with a linear projection of the decoder hidden state.
+        self._output_generation_layer = Linear(self.decoder_output_dim, target_vocab_size)
+
+        # We create a "copying" score for each source token by applying a non-linearity
+        # (tanh) to a linear projection of the encoded hidden state for that token,
+        # and then taking the dot product of the result with the decoder hidden state.
+        self._output_copying_layer = Linear(self.encoder_output_dim, self.decoder_output_dim)
+
+        # At prediction time, we'll use a beam search to find the best target sequence.
+        self._beam_search = BeamSearch(self._end_index, max_steps=max_decoding_steps, beam_size=beam_size)
+
+    @overrides
+    def forward(self,  # type: ignore
+                source_tokens: Dict[str, torch.LongTensor],
+                source_token_ids: torch.Tensor,
+                source_to_target: torch.Tensor,
+                metadata: List[Dict[str, Any]],
+                target_tokens: Dict[str, torch.LongTensor] = None,
+                target_token_ids: torch.Tensor = None) -> Dict[str, torch.Tensor]:
+        # pylint: disable=arguments-differ
+        """
+        Make foward pass with decoder logic for producing the entire target sequence.
+
+        Parameters
+        ----------
+        source_tokens : ``Dict[str, torch.LongTensor]``, required
+            The output of `TextField.as_array()` applied on the source `TextField`. This will be
+            passed through a `TextFieldEmbedder` and then through an encoder.
+        source_token_ids : ``torch.Tensor``, required
+            Tensor containing IDs that indicate which source tokens match each other.
+            Has shape: `(batch_size, trimmed_source_length)`.
+        source_to_target : ``torch.Tensor``, required
+            Tensor containing vocab index of each source token with respect to the
+            target vocab namespace. Shape: `(batch_size, trimmed_source_length)`.
+        metadata : ``List[Dict[str, Any]]``, required
+            Metadata field that contains the original source tokens ('source_tokens')
+            and any other meta fields.
```

Comment:
You also require `target_tokens` in here when using token-based metrics, right?

Reply:
Correct. I'll add a comment for that